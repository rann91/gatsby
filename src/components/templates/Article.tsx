/* eslint-disable react/prop-types */
/* eslint-disable react/display-name */
/* @jsx jsx */
import { graphql, PageProps } from 'gatsby'
import GatsbyImage, { FluidObject } from 'gatsby-image'
import ReactMarkdown from 'react-markdown'
import { AspectRatio, Box, Container, Heading, jsx, Text } from 'theme-ui'
import { rem } from '../../gatsby-plugin-theme-ui/index'
import { Strapi_Article, File } from '../../typings/graphql'
import { Fragment } from 'react'

type Props = PageProps<{
  strapi: {
    article: Strapi_Article
  }
}>

const Article = ({
  data: {
    strapi: { article }
  }
}: Props) => (
  <Box as="article" py={[7, null, null, 8]}>
    <Container>
      <Box px={5} sx={{ mx: 'auto', maxWidth: rem(740) }}>
        <Heading as="h1" variant="h1">
          {article.title}
        </Heading>
        <Text as="time" variant="label" pt={2}>
          {new Date(article.createdAt).toLocaleDateString('en-US', {
            month: 'long',
            day: 'numeric',
            year: 'numeric'
          })}
        </Text>
      </Box>
      <Box px={5} py={[5, null, null, 7]}>
        <AspectRatio ratio={16 / 9} bg="bg">
          {article.image && (
            <GatsbyImage
              fluid={article.image.file?.childImageSharp?.fluid as FluidObject}
              alt={article.image.alternativeText || undefined}
            />
          )}
        </AspectRatio>
      </Box>
      <Box px={5} sx={{ mx: 'auto', maxWidth: rem(740) }}>
        <ReactMarkdown
          source={article.content || undefined}
          renderers={{
            paragraph(props) {
              if (props.children[0].type.name === 'image') {
                return <Fragment {...props} />
              }

              return <p {...props} />
            },
            image({ src, alt }: HTMLImageElement) {
              const name = src
                .replace('/uploads/', '')
                .replace(/\.(jpe?g|png|gif)/, '')

              if (!article.contentFiles) {
                return null
              }

              const file = ((article.contentFiles || []) as File[]).find(
                (contentFile: File) =>
                  contentFile.name && contentFile.name.indexOf(name) > -1
              )

              if (!file) {
                return null
              }

              return (
                <GatsbyImage
                  fluid={file.childImageSharp?.fluid as FluidObject}
                  alt={alt}
                  sx={{ mb: 5 }}
                />
              )
            }
          }}
          sx={{
            h1: {
              variant: 'text.h1',
              m: 0,
              pb: 2
            },
            h2: {
              variant: 'text.h2',
              m: 0,
              pb: 2
            },
            h3: {
              variant: 'text.h3',
              m: 0,
              pb: 2
            },
            h4: {
              variant: 'text.h4',
              m: 0,
              pb: 2
            },
            h5: {
              variant: 'text.h5',
              m: 0,
              pb: 2
            },
            h6: {
              variant: 'text.h6',
              m: 0,
              pb: 2
            },
            p: {
              variant: 'text.p',
              m: 0,
              pb: 5
            },
            a: {
              color: 'primary'
            },
            blockquote: {
              m: 0,
              mb: 5,
              borderLeftWidth: 5,
              borderLeftStyle: 'solid',
              borderLeftColor: 'quoteBorder',
              p: {
                variant: 'text.quote',
                pl: 5,
                pb: 0
              }
            },
            ul: {
              my: 0
            },
            li: {
              pb: 2
            },
            img: {
              display: 'block',
              maxWidth: '100%'
            }
          }}
        />
      </Box>
    </Container>
  </Box>
)

export const query = graphql`
  query Article($id: ID!) {
    strapi {
      article(id: $id) {
        id
        meta {
          ...Meta
        }
        title
        slug
        image {
          ...ImageFluid
        }
        content
        contentFiles {
          ...File
          name
          childImageSharp {
            fluid(maxWidth: 740) {
              ...GatsbyImageSharpFluid
            }
          }
        }
        createdAt
      }
    }
  }
`

export default Article
