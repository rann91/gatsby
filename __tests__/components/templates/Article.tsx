import React from 'react'
import { render } from '@testing-library/react'
import Article from '../../../src/components/templates/Article'
import data from '../../fixtures/article.json'
import siteMetadata from '../../fixtures/site-metadata.json'
import { useStaticQuery } from 'gatsby'

const props = {
  path: '/blog/10-quick-tips-about-blogging',
  navigate: 'ƒ navigate() {}',
  location: {
    pathname: '/blog/10-quick-tips-about-blogging',
    search: '',
    hash: '',
    href: 'http://localhost:8000/blog/10-quick-tips-about-blogging',
    origin: 'http://localhost:8000',
    protocol: 'http:',
    host: 'localhost:8000',
    hostname: 'localhost',
    port: '8000',
    state: '{key: "1598424543569"}',
    key: '1598424543569',
    action: 'PUSH'
  },
  pageResources: {
    component: 'ƒ ExportedComponent() {}',
    json: '{data: {…}, pageContext: {…}}',
    page: '{componentChunkName: "component---src-components-te…}',
    staticQueryResults: '{2780586362: {…}, 2805985199: {…}, 3168435165: {…},…}'
  },
  uri: '/blog/10-quick-tips-about-blogging',
  pageContext: {
    id: '5f1c4cd100c3ff0bfb1beea4'
  },
  pathContext: {
    id: '5f1c4cd100c3ff0bfb1beea4'
  },
  data
}

beforeEach(() => {
  ;(useStaticQuery as jest.Mock).mockImplementationOnce(() => siteMetadata)
})

describe('Article', () => {
  it('renders correctly', () => {
    // @ts-ignore
    const { container } = render(<Article {...props} />)

    expect(container).toMatchSnapshot()
  })
})
