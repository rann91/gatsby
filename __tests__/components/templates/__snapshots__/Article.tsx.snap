// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article renders correctly 1`] = `
<div>
  <article
    class="css-1uy1wmc"
  >
    <div
      class="css-1yrmf69"
    >
      <div
        class="css-sia4fg-Article"
      >
        <h1
          class="css-1wf7gym"
        >
          10 Quick Tips About Blogging
        </h1>
        <time
          class="css-iwcjxb"
        >
          July 25, 2020
        </time>
      </div>
      <div
        class="css-b8zr0"
      >
        <div
          class="css-i9l3dq"
        >
          <div
            class="css-r01jhj"
          />
          <div
            class="css-1jyusi9"
          >
            <div
              class=" gatsby-image-wrapper"
              style="position: relative; overflow: hidden;"
            >
              <div
                aria-hidden="true"
                style="width: 100%; padding-bottom: 66.66666666666667%;"
              />
              <img
                alt=""
                aria-hidden="true"
                src="data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/2wBDARESEhgVGC8aGi9jQjhCY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2P/wgARCAANABQDASIAAhEBAxEB/8QAGQAAAgMBAAAAAAAAAAAAAAAAAAQBAgMF/8QAFQEBAQAAAAAAAAAAAAAAAAAAAAH/2gAMAwEAAhADEAAAAaa8hyIERf/EABoQAAIDAQEAAAAAAAAAAAAAAAECAAMSBBP/2gAIAQEAAQUCPOXmbM+hSK5EbqszP//EABURAQEAAAAAAAAAAAAAAAAAAAAR/9oACAEDAQE/AVf/xAAVEQEBAAAAAAAAAAAAAAAAAAAAEf/aAAgBAgEBPwFH/8QAHRAAAAYDAQAAAAAAAAAAAAAAAAECEBESISJxof/aAAgBAQAGPwKbkNIVwVUWWr63/8QAGxAAAgMBAQEAAAAAAAAAAAAAAREAITFhUZH/2gAIAQEAAT8h2+gm6hVCAC4+zzruQltwu4DgXHP/2gAMAwEAAgADAAAAEAQP/8QAFxEBAQEBAAAAAAAAAAAAAAAAAQARMf/aAAgBAwEBPxATtq//xAAXEQEBAQEAAAAAAAAAAAAAAAABABEx/9oACAECAQE/EEeWb//EABwQAQACAwEBAQAAAAAAAAAAAAERIQBBUXFhof/aAAgBAQABPxDWkQhgeXjmiTVFDSLxtruE/mCUl133uQI9qhD5zFSrKtq7z//Z"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 1; transition-delay: 500ms;"
              />
              <picture>
                <source
                  sizes="(max-width: 1920px) 100vw, 1920px"
                  srcset="/static/a328e26e48bf3ac82340bfbea4a5ce37/9104c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 480w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/a6352/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 960w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 1920w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/af56c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 2880w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/27fce/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 3840w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/947df/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 5472w"
                />
                <img
                  alt="Laptop and block note on a desk"
                  loading="lazy"
                  sizes="(max-width: 1920px) 100vw, 1920px"
                  src="/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg"
                  srcset="/static/a328e26e48bf3ac82340bfbea4a5ce37/9104c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 480w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/a6352/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 960w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 1920w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/af56c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 2880w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/27fce/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 3840w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/947df/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 5472w"
                  style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 0; transition: opacity 500ms;"
                />
              </picture>
              <noscript>
                &lt;picture&gt;&lt;source srcset="/static/a328e26e48bf3ac82340bfbea4a5ce37/9104c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 480w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/a6352/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 960w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 1920w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/af56c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 2880w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/27fce/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 3840w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/947df/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 5472w" sizes="(max-width: 1920px) 100vw, 1920px" /&gt;&lt;img loading="lazy" sizes="(max-width: 1920px) 100vw, 1920px" srcset="/static/a328e26e48bf3ac82340bfbea4a5ce37/9104c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 480w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/a6352/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 960w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 1920w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/af56c/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 2880w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/27fce/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 3840w,
/static/a328e26e48bf3ac82340bfbea4a5ce37/947df/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg 5472w" src="/static/a328e26e48bf3ac82340bfbea4a5ce37/d8255/laptop_desk_notebook_computer_writing_work_927799_pxhere_com_273164166b.jpg" alt="Laptop and block note on a desk" style="position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
              </noscript>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-sia4fg-Article"
      >
        <div
          class="css-cjphzs-Article"
        >
          <h2>
            Whatâ€™s a Rich Text element?
          </h2>
          <p>
            The rich text element allows you to create and format headings, paragraphs, blockquotes, images, and video all in one place instead of having to add and format them individually. Just double-click and easily create content.
          </p>
          <blockquote>
            <p>
              The rich text element allows you to create and format headings, paragraphs, blockquotes, images, and video all in one place instead of having to add and format them individually. Just double-click and easily create content.
            </p>
          </blockquote>
          <h3>
            Static and dynamic content editing
          </h3>
          <p>
            A rich text element can be used with static or dynamic content. For static content, just drop it into any page and begin editing. For dynamic content, add a rich text field to any collection and then connect a rich text element to that field in the settings panel. Voila!
          </p>
          <div
            class="css-19h5ycu gatsby-image-wrapper"
            style="position: relative; overflow: hidden;"
          >
            <div
              aria-hidden="true"
              style="width: 100%; padding-bottom: 74.5945945945946%;"
            />
            <img
              alt=""
              aria-hidden="true"
              src="data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/2wBDARESEhgVGC8aGi9jQjhCY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2P/wgARCAAPABQDASIAAhEBAxEB/8QAGAAAAgMAAAAAAAAAAAAAAAAAAAMCBAX/xAAWAQEBAQAAAAAAAAAAAAAAAAADAAH/2gAMAwEAAhADEAAAAXVyQNnjTb//xAAaEAEAAgMBAAAAAAAAAAAAAAABAhEAAxIx/9oACAEBAAEFAtnmwObvEnbFR4i//8QAFxEBAAMAAAAAAAAAAAAAAAAAAAEREv/aAAgBAwEBPwGKaf/EABYRAAMAAAAAAAAAAAAAAAAAAAABEf/aAAgBAgEBPwF1kP/EABgQAAMBAQAAAAAAAAAAAAAAAAAQIQEx/9oACAEBAAY/Aqqrp0//xAAZEAADAQEBAAAAAAAAAAAAAAAAAREhMUH/2gAIAQEAAT8hZSkqKsw0raSF3F6QBIwm1P/aAAwDAQACAAMAAAAQp/8A/8QAFhEBAQEAAAAAAAAAAAAAAAAAAQAR/9oACAEDAQE/ECA2zf/EABcRAAMBAAAAAAAAAAAAAAAAAAABESH/2gAIAQIBAT8QZmFn/8QAGxABAQADAQEBAAAAAAAAAAAAAREAIUExgZH/2gAIAQEAAT8QCJeUmn5l6bDwICzc7ikGJ64mWPRGpzAsrPAqZZpNad/M/9k="
              style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 1; transition-delay: 500ms;"
            />
            <picture>
              <source
                sizes="(max-width: 740px) 100vw, 740px"
                srcset="/static/70b0a429095ed777092409020ea8d4ad/29cff/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 185w,
/static/70b0a429095ed777092409020ea8d4ad/3b31f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 370w,
/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 740w,
/static/70b0a429095ed777092409020ea8d4ad/b3f9f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1110w,
/static/70b0a429095ed777092409020ea8d4ad/c6cd0/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1480w,
/static/70b0a429095ed777092409020ea8d4ad/73508/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 3359w"
              />
              <img
                alt="Red cat sitting"
                loading="lazy"
                sizes="(max-width: 740px) 100vw, 740px"
                src="/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg"
                srcset="/static/70b0a429095ed777092409020ea8d4ad/29cff/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 185w,
/static/70b0a429095ed777092409020ea8d4ad/3b31f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 370w,
/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 740w,
/static/70b0a429095ed777092409020ea8d4ad/b3f9f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1110w,
/static/70b0a429095ed777092409020ea8d4ad/c6cd0/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1480w,
/static/70b0a429095ed777092409020ea8d4ad/73508/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 3359w"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 0; transition: opacity 500ms;"
              />
            </picture>
            <noscript>
              &lt;picture&gt;&lt;source srcset="/static/70b0a429095ed777092409020ea8d4ad/29cff/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 185w,
/static/70b0a429095ed777092409020ea8d4ad/3b31f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 370w,
/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 740w,
/static/70b0a429095ed777092409020ea8d4ad/b3f9f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1110w,
/static/70b0a429095ed777092409020ea8d4ad/c6cd0/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1480w,
/static/70b0a429095ed777092409020ea8d4ad/73508/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 3359w" sizes="(max-width: 740px) 100vw, 740px" /&gt;&lt;img loading="lazy" sizes="(max-width: 740px) 100vw, 740px" srcset="/static/70b0a429095ed777092409020ea8d4ad/29cff/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 185w,
/static/70b0a429095ed777092409020ea8d4ad/3b31f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 370w,
/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 740w,
/static/70b0a429095ed777092409020ea8d4ad/b3f9f/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1110w,
/static/70b0a429095ed777092409020ea8d4ad/c6cd0/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 1480w,
/static/70b0a429095ed777092409020ea8d4ad/73508/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg 3359w" src="/static/70b0a429095ed777092409020ea8d4ad/83a78/sweet_animal_cute_pet_fur_red_675156_pxhere_com_6b56b8f673.jpg" alt="Red cat sitting" style="position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
            </noscript>
          </div>
          <h3>
            How to customize formatting for each rich text
          </h3>
          <p>
            Headings, paragraphs, blockquotes, figures, images, and figure captions can all be styled after a class is added to the rich text element using the "When inside of" nested selector system.
          </p>
          <ul>
            <li>
              The rich text element allows you to create and format headings
            </li>
            <li>
              Headings, paragraphs, blockquotes, figures, images, and figure captions
            </li>
            <li>
              A rich text element can be used with static or dynamic content
            </li>
          </ul>
        </div>
      </div>
    </div>
  </article>
</div>
`;
